// authenticate/authenticate.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'app',
				optional: true
			},
			{
				name: 'router',
				optional: true
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	//LOCAL FUNCTIONS ====================================



	//METHODS AND PROPERTIES ====================================

// app.use(require('serve-static')(__dirname + '/../../public'));
// app.use(require('cookie-parser')());
// ADDED to app in calling module: app.use(require('body-parser').urlencoded({ extended: true }));
// app.use(require('express-session')({ secret: 'keyboard cat', resave: true, saveUninitialized: true }));
// app.use(passport.initialize());
// app.use(passport.session());

	//INITIALIZATION ====================================
	
// app.post('/login', 
//   passport.authenticate('local', { failureRedirect: '/login' }),
//   function(req, res) {
//     res.redirect('/');
//   });	

this.router.post('/login', 
  function(req, res) {

if (req.body.userName=='tq' && req.body.password=='w'){
  
	res.json({
		userName:'tq',
		sessionKey:'admin:recipient',
		fullName:'TQ White II',
		roles:[]
	});
	}
	
	else{

			res.status('403').send('Bad authorization request');
	}
	
  });
  



	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
