// mz-web-user/mailing-list/mailing-list.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'app',
				optional: false
			},
			{
				name: 'router',
				optional: false
			},
			{
				name: 'userAccess',
				optional: false
			},
			{
				name: 'listAccess',
				optional: false
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};

	//LOCAL FUNCTIONS ====================================

	//METHODS AND PROPERTIES ====================================

	//INITIALIZATION ====================================

	this.router.post('/mailingList', function(req, res) {

		req.body.ownerEmailAdr = req.user.emailAdr; //prevent malicious outsiders from inserting owner who is not the logged in user

		self.listAccess.createList(req.body, function(err, result) {
			if (!err) {
				res.status('201').send(result);
			} else {
				res.status('403').send(err.message);
			}
		});

	});
	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
